File: Editorial/0000-0100/0022_Generate_Parentheses/Java/3_Divide_and_Conquer.java
class Solution {
    public List<String> generateParenthesis(int n) {
        if (n == 0) {
            return new ArrayList(Arrays.asList(""));
        }

        List<String> answer = new ArrayList();
        for (int leftCount = 0; leftCount < n; ++leftCount) {
            List<String> leftStrings = generateParenthesis(leftCount);
            List<String> rightStrings = generateParenthesis(n - 1 - leftCount);
            for (String leftString : leftStrings)
                for (String rightString : rightStrings)
                    answer.add("(" + leftString + ")" + rightString);
        }
        return answer;
    }
}

File: Editorial/0000-0100/0022_Generate_Parentheses/CSharp/3_Divide_and_Conquer.cs
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        if (n == 0) {
            return new List<string> { "" };
        }

        List<string> answer = new List<string>();
        for (int leftCount = 0; leftCount < n; ++leftCount) {
            IList<string> leftStrings = GenerateParenthesis(leftCount);
            IList<string> rightStrings = GenerateParenthesis(n - 1 - leftCount);
            foreach (string leftString in leftStrings) {
                foreach (string rightString in rightStrings) {
                    answer.Add("(" + leftString + ")" + rightString);
                }
            }
        }

        return answer;
    }
}
